Add servo tweening

ryftjet opened the pull request:
Added #5, moving the servo between the different levels smoothly.

ryftjet committed:
Make WiFi connection semi-asynchronous

Adjusted the `connect()` function so that it doesn't block the `loop()`.

ryftjet committed:
Merge pull request #9 from amberelectric/main

Add debugging to tweening development branch

ryftjet committed:
Add servo tweening for moving between levels

Used the `Tween` library to make the servo motor move smoothly between the different levels

madpilot commented on src/main.cpp:
```diff
@@ -204,6 +203,28 @@ String fetchDescriptor()
+  timeline.clear();
+  timeline.add(target);
+  if (target < SERVO_MIDPOINT) {
+    if (newTarget < SERVO_MIDPOINT && target > newTarget) {
```
Can you add a comment here about why there are two motion algorithms if the needle is either side of the midpoint?

madpilot replied:
I'm not sure you need all these if statements. Can't you just replace the whole thing with:

```cpp
if (newTarget > SERVO_MIDPOINT && target < newTarget) {
  timeline[target].then<Ease::BounceOut>(newTarget, abs(newTarget - target) * ANIMATION_SPEED);
} else {
  timeline[target].then<Ease::BackInOut>(newTarget, abs(newTarget - target) * ANIMATION_SPEED * 2);
}
```

`abs(newTarget - target)` and `abs(target - newTarget)` are equivalent aren't they?

ryftjet replied:
I don't think so. The current algorithm only uses bounce when going from high to moderate or extreme to catastrophic, but your algorithm uses bounce when going from (either extreme or catastrophic) to (either moderate or high). (I think)
<img width="1238" alt="Screenshot 2024-11-28 at 10 45 14" src="https://github.com/user-attachments/assets/2fe6e3e9-0199-4f99-9f6f-5bbbafa6e025">
I think it might be because the servo at 0˚ is as far right as it will go, and 180˚ is as far left as it will go. Not entirely sure, though.

madpilot commented on src/main.cpp:
```diff
@@ -215,24 +236,37 @@ void setup()
+    .then<Ease::BounceOut>(NO_DANGER, 2000);
+  timeline.start();
+
+  // Force a run 1500ms after startup (to allow servo animation to complete)
```
5000-300000 = -295000 - where are you getting 1500ms from?

ryftjet replied:
Yeah ok that's my bad. It was 1500ms but I changed it to 5000ms. As to where I'm getting the 5000 from,
`5000ms - 300000ms = -295000ms (-4min,55secs)`
So when the program starts, it sets the last API fetch time to 4 mins and 55 seconds ago. After 5 seconds, the elapsed time ticks over to 5 mins 0 seconds (300000ms), and it refreshes the API.

ryftjet committed:
Add better comments for decision making and math

Explain the decisions behind the servo animations, and the 5000ms API call delay at the start.
